src/SFML/Window/Unix/CursorImpl.cpp

```diff
@@ -0,0 +1,156 @@
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <SFML/Window/Unix/CursorImpl.hpp>
+#include <SFML/Window/Unix/Display.hpp>
+#include <X11/cursorfont.h>
+#include <X11/Xutil.h>
+#include <cassert>
+#include <cstdlib>
+#include <vector>
+
+namespace sf
+{
+namespace priv
+{
+
+////////////////////////////////////////////////////////////
+CursorImpl::CursorImpl() :
+m_display(OpenDisplay()),
+m_cursor(None)
+{
+    // That's it.
+}
+
+
+////////////////////////////////////////////////////////////
+CursorImpl::~CursorImpl()
+{
+    release();
+
+    CloseDisplay(m_display);
+}
+
+
+////////////////////////////////////////////////////////////
+bool CursorImpl::loadFromPixels(const Uint8* pixels, Vector2u size, Vector2u hotspot)
+{
+    release();
+
+    // Convert the image into a bitmap (monochrome!).
+    std::size_t bytes = (size.x + 7) / 8 * size.y;
+    std::vector<Uint8> mask(bytes, 0); // Defines which pixel is transparent.
+    std::vector<Uint8> data(bytes, 1); // Defines which pixel is white/black.
+
+    for (std::size_t j = 0; j < size.y; ++j)
+    {
+        for (std::size_t i = 0; i < size.x; ++i)
+        {
+            std::size_t pixelIndex = i + j * size.x;
+            std::size_t byteIndex  = pixelIndex / 8;
+            std::size_t bitIndex   = i % 8;
+
+            // Turn on pixel that are not transparent
+            Uint8 opacity = pixels[pixelIndex * 4 + 3] > 0 ? 1 : 0;
+            mask[byteIndex] |= opacity << bitIndex;
+
+            // Choose between black/background & white/foreground color for each pixel,
+            // based on the pixel color intensity: on average, if a channel is "active"
+            // at 25%, the bit is white.
+            int intensity = pixels[pixelIndex * 4 + 0] + pixels[pixelIndex * 4 + 1] + pixels[pixelIndex * 4 + 2];
+            Uint8 bit = intensity > 64 ? 1 : 0;
+            data[byteIndex] |= bit << bitIndex;
+        }
+    }
+
+    Pixmap maskPixmap = XCreateBitmapFromData(m_display, XDefaultRootWindow(m_display),
+                                              (char*)&mask[0], size.x, size.y);
+    Pixmap dataPixmap = XCreateBitmapFromData(m_display, XDefaultRootWindow(m_display),
+                                              (char*)&data[0], size.x, size.y);
+
+    // Define the foreground color as white and the background as black.
+    XColor fg, bg;
+    fg.red = fg.blue = fg.green = -1;
+    bg.red = bg.blue = bg.green =  0;
+
+    // Create the monochrome cursor.
+    m_cursor = XCreatePixmapCursor(m_display,
+                                   dataPixmap, maskPixmap,
+                                   &fg, &bg,
+                                   hotspot.x, hotspot.y);
+
+    // Free the resources
+    XFreePixmap(m_display, dataPixmap);
+    XFreePixmap(m_display, maskPixmap);
+
+    // We assume everything went fine...
+    return true;
```

Comment:
X11 being async, I think we simply cannot check for failure. Maybe some X11-guru can confirm that, but I've found nothing in the doc while writing this code to catche errors, sadly.